Refactor of hwctrl

BBScholar opened the pull request:
Adding new functionality is difficult. 

### Goals:
- Clean up code. 
- Redesign sensor hierarchy to make it easier to add sensors and such 
- Similar Abstraction for motors?

### Todo:
- Need to finish the refactor
- Need to test

edf42001 commented:
Hi Ben you seem to be refactoring some hwctrl, so just wanted to make sure you know about this other branch I'm merging https://github.com/cwruRobotics/NASA-RMC-2020/pull/28/files.

Hopefully doesn't conflict with yours when you're done

BBScholar commented:
Yup! I'm keeping an eye on it, and will make sure any changes I make will keep up with @Rjames13's work on the topics.

BBScholar commented:
At the moment at least, I don't think there will be any significant conflicts.

Rjames13 commented:
You are not really going to worry about me, but I know Jim is going to start doing the motor topics soon.

BBScholar commented:
Do not merge this yet, it needs to be tested and reviewed.

edf42001 commented:
Flattered that you think I know what's going on here but you'll probably just want Adam to look at it 

Although I'm definitely curious.

BBScholar commented:
## Overview
Take a look at this when you can. It isn't 100% done yet, but most functionality is there. My c++ skills are a bit rusty so check for stupid mistakes if you have the time. Going to delete the old hwctrl (in this branch), and rename this one to hwctrl. I'lll keep this up to date with what @Rjames13 and @JHorwitz1011 are doing as well.

Changes:
- Made motor and sensor abstraction less messy (using virtual functions, templates, etc).
- Got rid of a lot of c style code to allow easier maintenance. 
- increased utilization of boost libs, std lib, and smart pointers
- Changed how we implement new threads in the controller
- implemented the quad encoder sensor (mostly)
- implemented the bmc (almos, still a work in progress)
- Made calibration a seperate file
- organized files

Issues:
- Compile time will probably be longer.

BBScholar commented:
> Flattered that you think I know what's going on here but you'll probably just want Adam to look at it
> 
> Although I'm definitely curious.

haha. Got it.

edf42001 commented:
No offense but I'm totally unsubscribing from these commit notifications (@ anyone else you can do that do if it's annoying you)

edf42001 commented:
Also is it called hwctrl2 because we're keeping the other hwctrl in the code for now?

BBScholar commented:
> Also is it called hwctrl2 because we're keeping the other hwctrl in the code for now?

No, It was mainly so I can copy code from the 1st version without opening a browser or switching branches. I deleted the old one in my branch, and renamed it to hwctrl. And no offense taken on the github notifications haha.

BBScholar commented:
Compilation takes 24 minutes with optimization.

BBScholar commented:
Beaglebone is basically a potato

edf42001 commented:
I see. That sounds like a problem

edf42001 commented:
What if we crossed compiled it

BBScholar commented:
We've actually been talking about that. It just seems like a bit of a pain to get it working with cmake. I'll look into it a bit more.

edf42001 commented:
I know it's the middle of winter break, but I was hit with a sudden curiosity to know when this will be ready to be merged?

BBScholar commented:
@edf42001 Very soon. Adam wanted me to fix one thing related to the vesc controllers. Should be good after his review. I still need to finish up some implementations of certain sensors, but I'm doing that as I work on their firmware.

AdamC19 reviewed (approved):
I think the goals of this refactoring have been completed. The code is more organized and easier to build on.
